apps/prairielearn/src/pages/elementFiles/elementFiles.js

```diff
@@ -24,67 +23,82 @@ const sql = sqldb.loadSqlEquiv(import.meta.url);
 
-    // If the route includes a `cachebuster` param, we'll set the `immutable`
-    // and `maxAge` options on the `Cache-Control` header. This router is
-    // mounted twice - one with the cachebuster in the URL, and once without it
-    // for backwards compatibility. See `server.js` for more details.
-    //
-    // As with `/assets/`, we assume that element files are likely to change
-    // when running in dev mode, so we skip caching entirely in that case.
-    const isCached = !!req.params.cachebuster && !config.devMode;
-    const sendFileOptions = {
-      immutable: isCached,
-      maxAge: isCached ? '31536000s' : 0,
-    };
+      // If the route includes a `cachebuster` param, we'll set the `immutable`
+      // and `maxAge` options on the `Cache-Control` header. This router is
+      // mounted twice - one with the cachebuster in the URL, and once without it
+      // for backwards compatibility. See `server.js` for more details.
+      //
+      // As with `/assets/`, we assume that element files are likely to change
+      // when running in dev mode, so we skip caching entirely in that case.
+      const isCached = !!req.params.cachebuster && !config.devMode;
+      const sendFileOptions = {
+        immutable: isCached,
+        maxAge: isCached ? '31536000s' : 0,
+      };
 
-    if (isCached) {
-      // `middlewares/cors.js` disables caching for all routes by default.
-      // We need to remove this header so that `res.sendFile` can set it
-      // correctly.
-      res.removeHeader('Cache-Control');
-    }
+      if (isCached) {
+        // `middlewares/cors.js` disables caching for all routes by default.
+        // We need to remove this header so that `res.sendFile` can set it
+        // correctly.
+        res.removeHeader('Cache-Control');
+      }
 
-    let elementFilesDir;
-    if (res.locals.course) {
-      // Files should be served from the course directory
-      let question_course;
-      if (req.params.producing_course_id) {
-        const producing_course_id = z.string().parse(req.params.producing_course_id);
-        const has_shared_question = await sqldb.queryRow(
-          sql.select_has_shared_question,
-          { consuming_course_id: res.locals.course.id, producing_course_id },
+      let elementFilesDir;
+      if (options.publicEndpoint && !options.static) {
+        const has_publicy_shared_question = await sqldb.queryRow(
+          sql.select_has_publicly_shared_question,
+          { course_id: req.params.course_id },
           z.boolean(),
         );
-        if (!has_shared_question) {
+        if (!has_publicy_shared_question) {
           throw new HttpStatusError(404, 'Not Found');
         }
+        const course = await selectCourseById(req.params.course_id);
+        const coursePath = chunks.getRuntimeDirectoryForCourse(course);
+        await chunks.ensureChunksForCourseAsync(course.id, { type: 'elements' });
+
+        elementFilesDir = path.join(coursePath, 'elements');
```

Comment:
This is the only actual new code here, the rest of the diff is muddled by the indentation level change.